src/dbnode/storage/namespace/context.go

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package namespace
+
+import (
+	"github.com/m3db/m3/src/x/ident"
+)
+
+type Context struct {
+	Id     ident.ID
+	Schema SchemaDescr
+}
+
+func NewContextFrom(nsMetadata Metadata) Context {
+	ctx := Context{Id: nsMetadata.ID()}
+	if schema, ok := nsMetadata.Options().SchemaHistory().GetLatest(); ok {
+		ctx.Schema = schema
+	}
+	return ctx
+}
+
+func NewContextFor(id ident.ID, registry SchemaRegistry) Context {
+	ctx := Context{Id: id}
+	if schema, ok := registry.GetLatestSchema(id); ok {
```

Comment:
Yeah this is going to be a very high frequency call, can we cache the schema on the namespace and pass that along in `WriteOptions` please?

Let's discuss this further if you have differing opinions, but adding a second RLock here and a map lookup and a string conversion to each call to `Write(..)` for the hot path is quite a significant addition and likely quite hard to undo later.

Reply:
I am all for avoiding additional lock for hot path.

how long do we have to cache schema for? is it still for the duration of a single write?